src/include/daos/tls.h

```diff
@@ -0,0 +1,124 @@
+/**
+ * (C) Copyright 2016-2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * This file is part of daos
+ *
+ * src/include/daos/tls.h
+ */
+
+#ifndef __DAOS_TLS_H__
+#define __DAOS_TLS_H__
+
+#include <daos/common.h>
+#include <daos_types.h>
+
+/**
+ * Stackable Module API
+ * Provides a modular interface to load and register server-side code on
+ * demand. A module is composed of:
+ * - a set of request handlers which are registered when the module is loaded.
+ * - a server-side API (see header files suffixed by "_srv") used for
+ *   inter-module direct calls.
+ *
+ * For now, all loaded modules are assumed to be trustful, but sandboxes can be
+ * implemented in the future.
+ */
+/*
+ * Thead-local storage
+ */
+struct daos_thread_local_storage {
+	uint32_t	dtls_tag;
+	void		**dtls_values;
+};
+
+enum daos_module_tag {
+	DAOS_SYS_TAG    = 1 << 0, /** only run on system xstream */
+	DAOS_TGT_TAG    = 1 << 1, /** only run on target xstream */
+	DAOS_RDB_TAG    = 1 << 2, /** only run on rdb xstream */
+	DAOS_OFF_TAG    = 1 << 3, /** only run on offload/helper xstream */
+	DAOS_CLI_TAG    = 1 << 4, /** only run on client stack */
+	DAOS_SERVER_TAG = 0xff,   /** run on all xstream */
+};
+
+/* The module key descriptor for each xstream */
+struct daos_module_key {
+	/* Indicate where the keys should be instantiated */
+	enum daos_module_tag dmk_tags;
+
+	/* The position inside the daos_module_keys */
+	int dmk_index;
+	/* init keys for context */
+	void *(*dmk_init)(int tags, int xs_id, int tgt_id);
+
+	/* fini keys for context */
+	void (*dmk_fini)(int tags, void *data);
+};
+
+#define DAOS_MODULE_KEYS_NR 10
+extern pthread_key_t dc_tls_key;	/* client side TLS key */
+extern pthread_key_t dss_tls_key;	/* server side TLS key */
+extern struct daos_module_key *daos_module_keys[DAOS_MODULE_KEYS_NR];
+
+static inline struct daos_thread_local_storage *
+dss_tls_get()
+{
+	return (struct daos_thread_local_storage *)
+		pthread_getspecific(dss_tls_key);
+}
+
+static inline struct daos_thread_local_storage *
+dc_tls_get()
+{
+	return (struct daos_thread_local_storage *)
+		pthread_getspecific(dc_tls_key);
+}
+
+/* For now TLS is only enabled if metrics are enabled */
+#define METRIC_DUMP_ENV "DAOS_METRIC_DUMP_ENV"
+#define DAOS_CLIENT_METRICS_ENV	"DAOS_CLIENT_METRICS"
+#define DAOS_CLIENT_METRICS_RETAIN_ENV	"DAOS_CLIENT_METRICS_RETAIN"
+extern bool client_metric;
```

Comment:
You can't have a variable called "client_metric" in a shared library that you're linking with end-user applications.

Reply:
sure